Crashlytics/Crashlytics/Models/FIRCLSOnDemandModel.m

```diff
@@ -0,0 +1,202 @@
+ * rate limiting and exponential backoff.
+ */
+- (BOOL)recordOnDemandExceptionIfQuota:(FIRExceptionModel *)exceptionModel
+             withDataCollectionEnabled:(BOOL)dataCollectionEnabled
+            usingExistingReportManager:(FIRCLSExistingReportManager *)existingReportManager {
+  // Record the exception model into a new report if there is unused on-demand quota. Otherwise,
+  // log the occurence but drop the event.
+  @synchronized(self) {
+    if (![self isQueueFull]) {
+      FIRCLSDataCollectionToken *dataCollectionToken = [FIRCLSDataCollectionToken validToken];
+      NSString *activeReportPath = FIRCLSExceptionRecordOnDemandModel(
+          exceptionModel, self.recordedOnDemandExceptionCount, self.droppedOnDemandExceptionCount);
+
+      // Only submit an exception report if data collection is enabled. Otherwise, the report
+      // is stored until send or delete unsent reports is called.
+      if (dataCollectionEnabled) {
+        if (activeReportPath) {
+          [self incrementQueuedOperationCount:1];
+          [self getOrIncrementOnDemandEventCountForCurrentRun:YES];
+          dispatch_async(self.dispatchQueue, ^{
+            [existingReportManager handleOnDemandReportUpload:activeReportPath
+                                          dataCollectionToken:dataCollectionToken
+                                                     asUrgent:YES];
+            double uploadDelay = [self calculateUploadDelay];
+            FIRCLSInfoLog(@"Submitted an on-demand exception, starting delay %.20f", uploadDelay);
+            sleep(uploadDelay);
+          });
+          return YES;  // Recorded and submitted the exception.
+        } else {
+          FIRCLSInfoLog(@"Error recording on-demand exception");
+          return NO;  // Something went wrong when recording the exception, so we don't have a valid
+                      // path.
+        }
+      }
+      // TODO: should we update delay if data collection is off?
+      // TODO: should we update counter(s)?
+      //      FIRCLSInfoLog(@"Submitted an on-demand exception with delay %d", uploadDelay);
+      //      [self updateBucketsAndCountWithDelay:uploadDelay];
+      return YES;  // Recorded exception but did not submit since data collection is disabled.
+    } else {
+      FIRCLSInfoLog(@"No available on-demand quota. Dropping report.");
+      [self getOrIncrementDroppedOnDemandEventCountForCurrentRun:YES];
+      return NO;  // Didn't record or submit the exception because no quota was available.
+    }
+  }
+}
+
+// Helper methods
+
+- (double)calculateUploadDelay {
+  double calculatedStepDuration = [self calculateStepDuration];
+  double power = pow(self.base, calculatedStepDuration);
+  NSNumber *calculatedUploadDelay =
+      [NSNumber numberWithDouble:(SEC_PER_MINUTE / self.uploadRate) * power];
+  NSComparisonResult result =
+      [[NSNumber numberWithDouble:MAX_DELAY_SEC] compare:calculatedUploadDelay];
+  [self incrementQueuedOperationCount:-1];
```

Comment:
This seems hidden in this method based on its name - I would expect `calculateUploadDelay` to be pure.  Can we call this in the dispatch_async block?